PTMCMCSampler/PTMCMCSampler.py

```diff
@@ -601,92 +598,46 @@ def PTswap(self, p0, lnlike0, lnprob0, iter):
         @return lnlike0: new log-likelihood
         @return lnprob0: new log posterior value
 
+        Repurposed from Neil Cornish/Bence Becsy's code:
         """
+        Ts = self.ladder
 
-        # initialize variables
-        readyToSwap = 0
-        swapAccepted = 0
-        swapProposed = 0
-
-        # if Tskip is reached, block until next chain in ladder is ready for
-        # swap proposal
-        if iter % self.Tskip == 0 and self.MPIrank < self.nchain - 1:
-            swapProposed = 1
-
-            # send current likelihood for swap proposal
-            self.comm.send(lnlike0, dest=self.MPIrank + 1, tag=18)
-
-            # determine if swap was accepted
-            swapAccepted = self.comm.recv(source=self.MPIrank + 1, tag=888)
+        log_Ls = self.comm.gather(lnlike0, root=0)  # list of likelihoods from each chain
+        p0s = self.comm.gather(p0, root=0)  # list of parameter arrays from each chain
 
-            # perform swap
-            if swapAccepted:
-
-                # exchange likelihood
-                lnlike0 = self.comm.recv(source=self.MPIrank + 1, tag=18)
-
-                # exchange parameters
-                pnew = np.empty(self.ndim)
-                self.comm.Sendrecv(
-                    p0, dest=self.MPIrank + 1, sendtag=19, recvbuf=pnew, source=self.MPIrank + 1, recvtag=19
-                )
-                p0 = pnew
-
-                # calculate new posterior values
-                lnprob0 = 1 / self.temp * lnlike0 + self.logp(p0)
-
-        # check if next lowest temperature is ready to swap
-        elif self.MPIrank > 0:
-
-            readyToSwap = self.comm.Iprobe(source=self.MPIrank - 1, tag=18)
-            # trick to get around processor using 100% cpu while waiting
-            time.sleep(0.000001)
-
-            # hotter chain decides acceptance
-            if readyToSwap:
-                newlnlike = self.comm.recv(source=self.MPIrank - 1, tag=18)
-
-                # determine if swap is accepted and tell other chain
-                logChainSwap = (1 / self.ladder[self.MPIrank - 1] - 1 / self.ladder[self.MPIrank]) * (
-                    lnlike0 - newlnlike
-                )
-
-                if logChainSwap > np.log(np.random.rand()):
-                    swapAccepted = 1
+        if self.MPIrank == 0:
```

Comment:
This block looks a lot simpler (and so probably more maintainable / less error-prone) to me than the iteration on the older code I made in #30.  It does introduce a dependency on the rank 0 process for all swaps, so there's a chance the code in #30 would perform better at large scale since more work could happen in parallel (and no potential communication bottleneck would exist here for rank 0 exchanging data with other processes).  I'm relying on the previous review to for correctness.

Reply:
The motivation to lock everything to the rank 0, cold chain is to fix a problem that occurs when different parts of parameter space have different compute times for the posterior.  This effectively breaks detailed balance as PT swaps are more likely to propose moves into slow compute regions.